commandsv3/src/main/java/org/wpilib/commands3/sysid/SysIdRoutine.java

```diff
@@ -0,0 +1,289 @@
+     *     routines.
+     * @param log Returns measured data of the mechanism motors during test routines. To return
+     *     data, call `motor(string motorName)` on the supplied `SysIdRoutineLog` instance, and then
+     *     call one or more of the chainable logging handles (e.g. `voltage`) on the returned
+     *     `MotorLog`. Multiple motors can be logged in a single callback by calling `motor`
+     *     multiple times.
+     * @param subsystem The subsystem containing the motor(s) that is (or are) being characterized.
+     *     Will be declared as a requirement for the returned test commands. The subsystem's `name`
+     *     will be appended to the log entry title for the routine's test state, e.g.
+     *     "sysid-test-state-subsystem".
+     */
+    public SysIdMechanism(
+        Consumer<Voltage> drive, Consumer<SysIdRoutineLog> log, Mechanism subsystem) {
+      this(drive, log, subsystem, null);
+    }
+  }
+
+  /** Motor direction for a SysId test. */
+  public enum Direction {
+    /** Forward. */
+    kForward,
+    /** Reverse. */
+    kReverse
+  }
+
+  /**
+   * Returns a command to run a quasistatic test in the specified direction.
+   *
+   * <p>The command will call the `drive` and `log` callbacks supplied at routine construction once
+   * per iteration. Upon command end or interruption, the `drive` callback is called with a value of
+   * 0 volts.
+   *
+   * @param direction The direction in which to run the test.
+   * @return A command to run the test.
+   */
+  public Command quasistatic(Direction direction) {
+    State state;
+    if (direction == Direction.kForward) {
+      state = State.kQuasistaticForward;
+    } else { // if (direction == Direction.kReverse) {
+      state = State.kQuasistaticReverse;
+    }
+
+    double outputSign = direction == Direction.kForward ? 1.0 : -1.0;
+
+    Timer timer = new Timer();
+    return Command.sequence(
+            m_mechanism.m_subsystem.run(co -> timer.restart()).named("sysid-Restart Timer"),
+            m_mechanism
+                .m_subsystem
+                .run(
+                    co -> {
+                      while (true) {
+                        m_mechanism.m_drive.accept(
+                            (Voltage)
+                                m_config.m_rampRate.times(Seconds.of(timer.get() * outputSign)));
+                        m_mechanism.m_log.accept(this);
+                        m_recordState.accept(state);
+                        co.yield();
+                      }
+                    })
+                .whenCanceled(
+                    () -> {
+                      m_mechanism.m_drive.accept(Volts.of(0));
+                      m_recordState.accept(State.kNone);
+                      timer.stop();
+                    })
+                .named("sysid-run quasistatic test"))
+        .named("sysid-" + state.toString() + "-" + m_mechanism.m_name)
+        .withTimeout(m_config.m_timeout);
```

Comment:
This could be a bit more idiomatic:

```suggestion
    return m_mechanism.m_subsystem.run(co -> {
      var timer = new Timer();
      timer.start();
      while (!timer.hasElapsed(m_config.m_timeout)) {
        var voltage = (Voltage) m_config.m_rampRate.times(Seconds.of(timer.get() * outputSign));
        m_mechanism.m_drive.accept(voltage);
        m_mechanism.m_log.accept(SysIdRoutine.this);
        m_recordState.accept(state);
        co.yield();
      }
      m_mechanism.m_drive.accept(Volts.zero());
      m_recordState.accept(State.kNone);
    })
    .whenCanceled(() -> m_mechanism.m_drive.accept(Volts.zero()))
    .named("sysid-" + state + "-" + m_mechanism.m_name);
```